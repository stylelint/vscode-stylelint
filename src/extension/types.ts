import type TypedEventEmitter from 'typed-emitter';
import type { DidRegisterDocumentFormattingEditProviderNotificationParams } from '../server/index';

/**
 * VS Code extension event names.
 */
export type ExtensionEvents = {
	DidRegisterDocumentFormattingEditProvider: (
		params: DidRegisterDocumentFormattingEditProviderNotificationParams,
	) => void;
	DidResetConfiguration: () => void;
};

/**
 * VS Code extension public API.
 */
export type PublicApi = TypedEventEmitter<ExtensionEvents> & {
	codeActionReady: boolean;
};

/**
 * Extension API event names.
 */
export enum ApiEvent {
	DidRegisterDocumentFormattingEditProvider = 'DidRegisterDocumentFormattingEditProvider',
	DidResetConfiguration = 'DidResetConfiguration',
}
