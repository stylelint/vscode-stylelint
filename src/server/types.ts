import type { Connection, Disposable } from 'vscode-languageserver';
import type { TextDocuments } from 'vscode-languageserver/node';
import type { TextDocument, TextEdit } from 'vscode-languageserver-textdocument';
import { CodeActionKind as VSCodeActionKind } from 'vscode-languageserver-types';
import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line n/no-unpublished-import
import type stylelint from 'stylelint';
import type winston from 'winston';
import type { StylelintRunner, LintDiagnostics } from '../utils/stylelint/index';
import type { ExtractKeysOfValueType } from '../utils/types';
import type { PackageManager, StylelintResolutionResult } from '../utils/packages/index';
import type { NotificationManager, CommandManager } from '../utils/lsp/index';

/**
 * Command IDs
 */
export enum CommandId {
	ApplyAutoFix = 'stylelint.applyAutoFix',
	OpenRuleDoc = 'stylelint.openRuleDoc',
}

/**
 * Code action kinds
 */
export const CodeActionKind = {
	StylelintSourceFixAll: `${VSCodeActionKind.SourceFixAll}.stylelint`,
};

/**
 * Language server notification names.
 */
export enum Notification {
	DidRegisterCodeActionRequestHandler = 'stylelint/didRegisterCodeActionRequestHandler',
	DidRegisterDocumentFormattingEditProvider = 'stylelint/didRegisterDocumentFormattingEditProvider',
	DidResetConfiguration = 'stylelint/didResetConfiguration',
}

/**
 * `DidRegisterDocumentFormattingEditProvider` notification parameters.
 */
export interface DidRegisterDocumentFormattingEditProviderNotificationParams {
	/**
	 * The URI of the document for which the formatting provider was registered.
	 */
	readonly uri: string;

	/**
	 * The options used to register the document formatting provider.
	 */
	readonly options: LSP.DocumentFormattingRegistrationOptions;
}

/**
 * Context shared between the language server and its modules.
 */
export interface LanguageServerContext {
	/**
	 * The language server connection.
	 */
	connection: Connection;

	/**
	 * The notification manager for the connection.
	 */
	notifications: NotificationManager;

	/**
	 * The command manager for the connection.
	 */
	commands: CommandManager;

	/**
	 * The text document manager.
	 */
	documents: TextDocuments<TextDocument>;

	/**
	 * The runner with which to run Stylelint.
	 */
	runner: StylelintRunner;

	/**
	 * Displays the given error in the client using the language server
	 * connection.
	 * @param error The error to display.
	 */
	displayError(error: unknown): void;

	/**
	 * Gets the effective extension options for a resource, given its URI.
	 * @param uri The resource URI.
	 */
	getOptions(uri: string): Promise<LanguageServerOptions>;

	/**
	 * Returns the module with the given ID if it exists.
	 * @param id The ID of the module to return.
	 */
	getModule(id: string): LanguageServerModule | undefined;

	/**
	 * Lints a document using Stylelint and returns fix text edits.
	 * @param document The document to get text edits for.
	 * @param linterOptions Options to pass to the linter. Overridden by the
	 * language server options.
	 */
	getFixes(document: TextDocument, linterOptions?: stylelint.LinterOptions): Promise<TextEdit[]>;

	/**
	 * Lints a document using Stylelint and returns diagnostics.
	 * @param document The document to lint.
	 * @param linterOptions Options to pass to the linter. Overridden by the
	 * language server options.
	 */
	lintDocument(
		document: TextDocument,
		linterOptions?: Partial<stylelint.LinterOptions>,
	): Promise<LintDiagnostics | undefined>;

	/**
	 * Resolves the Stylelint package to be used for the given document.
	 * @param document The document to resolve the package for.
	 */
	resolveStylelint(document: TextDocument): Promise<StylelintResolutionResult | undefined>;
}

/**
 * A language server module.
 */
export interface LanguageServerModule extends Disposable {
	/**
	 * Handler called when the server is initializing.
	 */
	onInitialize?: (params: LSP.InitializeParams) => Partial<LSP.InitializeResult> | undefined | void;

	/**
	 * Handler called after the language server finishes registering handlers
	 * with the connection.
	 */
	onDidRegisterHandlers?: () => void;

	/**
	 * Handler called after the language server has finished responding to the
	 * onDidChangeConfiguration event.
	 */
	onDidChangeConfiguration?: () => Promise<void>;

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	[key: string | symbol]: any;
}

/**
 * A language server module class.
 */
export interface LanguageServerModuleConstructor {
	/**
	 * The module's ID, used to identify the module in the language server's
	 * internal state and when logging. Should be a unique, short, lowercase
	 * string.
	 */
	id: string;
	new (params: LanguageServerModuleConstructorParameters): LanguageServerModule;
}

/**
 * Parameters for the {@link LanguageServerModuleConstructor} constructor.
 */
export type LanguageServerModuleConstructorParameters = {
	context: LanguageServerContext;
	logger?: winston.Logger;
};

/**
 * Language server event handler names.
 */
export type LanguageServerHandlers = ExtractKeysOfValueType<LanguageServerModule, () => unknown>;

/**
 * Parameters for language server event handlers, keyed by the handler name.
 */
export type LanguageServerHandlerParameters = {
	[key in LanguageServerHandlers]: Parameters<Required<LanguageServerModule>[key]>;
};

/**
 * Return types for language server event handlers, keyed by the handler name.
 */
export type LanguageServerHandlerReturnValues = {
	[key in LanguageServerHandlers]: ReturnType<Required<LanguageServerModule>[key]>;
};

/**
 * Language server constructor parameters.
 */
export type LanguageServerConstructorParameters = {
	/**
	 * The language server connection.
	 */
	connection: Connection;

	/**
	 * The logger to use. If not provided, a default logger will be used.
	 */
	logger?: winston.Logger;

	/**
	 * The modules to load.
	 */
	modules?: LanguageServerModuleConstructor[];
};

/**
 * Language server options.
 */
export type LanguageServerOptions = {
	codeAction: {
		disableRuleComment: {
			location: 'separateLine' | 'sameLine';
		};
	};
	config?: stylelint.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager: PackageManager;
	reportDescriptionlessDisables?: boolean;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet: string[];
	stylelintPath?: string;
	validate: string[];
};
