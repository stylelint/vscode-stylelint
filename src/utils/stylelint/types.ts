import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line n/no-unpublished-import
import type stylelint from 'stylelint';
import type { PackageManager } from '../packages/index';
export type Stylelint = typeof stylelint;
export type ConfigurationError = Error & { code: 78 };

/**
 * Diagnostics for a lint run.
 */
export type LintDiagnostics = {
	/**
	 * The diagnostics, each corresponding to a warning or error emitted by
	 * Stylelint.
	 */
	diagnostics: LSP.Diagnostic[];

	/**
	 * Raw output from Stylelint, if any.
	 */
	output?: string;

	/**
	 * Gets the original warning from the given diagnostic.
	 */
	getWarning?: (diagnostic: LSP.Diagnostic) => stylelint.Warning | null;
};

/**
 * Disable report rule names.
 */
export enum DisableReportRuleNames {
	Needless = '--report-needless-disables',
	InvalidScope = '--report-invalid-scope-disables',
	Descriptionless = '--report-descriptionless-disables',
	Illegal = 'reportDisables',
}

/**
 * Stylelint runner options.
 */
export type RunnerOptions = {
	config?: stylelint.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager?: PackageManager;
	reportDescriptionlessDisables?: boolean;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet?: string[];
	stylelintPath?: string;
	validate?: string[];
};

/**
 * Error thrown when a rule's option is invalid.
 */
export class InvalidOptionError extends Error {
	reasons: string[];

	constructor(warnings: { text: string }[]) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}
