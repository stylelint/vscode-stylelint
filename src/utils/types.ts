/**
 * Makes all properties of type `T` optional except for those in `K`.
 */
export type OptionalExcept<T extends Record<never, never>, K extends keyof T> = Pick<T, K> &
	Partial<Pick<T, K>>;

/**
 * Makes all properties of type `T` required except for those in `K`.
 */
export type RequiredExcept<T extends Record<never, never>, K extends keyof T> = Pick<
	T,
	Exclude<keyof T, K>
> &
	Required<Pick<T, K>>;

/**
 * Extracts keys from a type `T` the values of which are type `V`.
 */
export type ExtractKeysOfValueType<T, V> = {
	[K in keyof T]-?: T[K] extends V ? K : never;
}[keyof T];

/**
 * Given a function `T`, returns a function type with the same parameters, but
 * with the return type `Promise<R> | R` where `R` is the return type of `T`.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type MaybeAsync<T extends (...args: any[]) => any> = (
	...args: Parameters<T>
) => Promise<ReturnType<T>> | ReturnType<T>;

/**
 * Given a type `T`, returns a type with only the public properties of `T`.
 */
export type PublicOnly<T> = Pick<T, keyof T>;

/**
 * Given a type `T`, returns a type with only the public properties of `T` and
 * its descendants.
 */
export type PublicOnlyDeep<T> = {
	// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
	[K in keyof T]: T[K] extends Function ? T[K] : T[K] extends object ? PublicOnlyDeep<T[K]> : T[K];
};
