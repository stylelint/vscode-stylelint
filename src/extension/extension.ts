import { EventEmitter } from 'node:events';
import path from 'node:path';

import { LanguageClient, SettingMonitor, ExecuteCommandRequest } from 'vscode-languageclient/node';
import { workspace, commands, window, type ExtensionContext } from 'vscode';
import { ApiEvent, PublicApi } from './types';
import {
	CommandId,
	DidRegisterDocumentFormattingEditProviderNotificationParams,
	Notification,
} from '../server/index';

let client: LanguageClient;

/**
 * Activates the extension.
 */
export async function activate({ subscriptions }: ExtensionContext): Promise<PublicApi> {
	const serverPath = path.join(__dirname, 'start-server.js');

	const api = Object.assign(new EventEmitter(), {
		codeActionReady: false,
	}) as PublicApi;

	client = new LanguageClient(
		'Stylelint',
		{
			run: {
				module: serverPath,
			},
			debug: {
				module: serverPath,
				options: {
					execArgv: ['--nolazy', '--inspect=6004'],
				},
			},
		},
		{
			documentSelector: [{ scheme: 'file' }, { scheme: 'untitled' }],
			diagnosticCollectionName: 'Stylelint',
			synchronize: {
				fileEvents: [
					workspace.createFileSystemWatcher('**/.stylelintrc{,.js,.json,.yaml,.yml}'),
					workspace.createFileSystemWatcher('**/{stylelint.config.js,.stylelintignore}'),
				],
			},
		},
	);

	const errorHandler = async (error: unknown): Promise<void> => {
		await window.showErrorMessage(
			`Stylelint: ${error instanceof Error ? error.message : String(error)}`,
		);
	};

	const notificationHandlers = (): void => {
		client.onNotification(Notification.DidRegisterCodeActionRequestHandler, () => {
			api.codeActionReady = true;
		});
		client.onNotification(
			Notification.DidRegisterDocumentFormattingEditProvider,
			(params: DidRegisterDocumentFormattingEditProviderNotificationParams) => {
				api.emit(ApiEvent.DidRegisterDocumentFormattingEditProvider, params);
			},
		);
		client.onNotification(Notification.DidResetConfiguration, () => {
			api.emit(ApiEvent.DidResetConfiguration);
		});
	};

	try {
		await client.start();
		notificationHandlers();
	} catch (err) {
		await errorHandler(err);
	}

	subscriptions.push(
		// cspell:disable-next-line
		commands.registerCommand('stylelint.executeAutofix', async () => {
			const textEditor = window.activeTextEditor;

			if (!textEditor) {
				return;
			}

			const textDocument = {
				uri: textEditor.document.uri.toString(),
				version: textEditor.document.version,
			};
			const params = {
				command: CommandId.ApplyAutoFix,
				arguments: [textDocument],
			};

			try {
				await client.sendRequest(ExecuteCommandRequest.type, params);
			} catch {
				await window.showErrorMessage(
					'Failed to apply Stylelint fixes to the document. Please consider opening an issue with steps to reproduce.',
				);
			}
		}),
	);

	subscriptions.push(
		commands.registerCommand('stylelint.restart', async () => {
			await client.stop();

			try {
				await client.start();
				notificationHandlers();
			} catch (error: unknown) {
				await errorHandler(error);
			}
		}),
	);

	subscriptions.push(new SettingMonitor(client, 'stylelint.enable').start());

	return Promise.resolve(api);
}

/**
 * @returns A promise that resolves when the client has been deactivated.
 */
export async function deactivate(): Promise<void> {
	if (client) {
		try {
			return client.stop();
		} catch (err) {
			const msg = err && (err as Error) ? (err as Error).message : 'unknown';

			await window.showErrorMessage(`error stopping stylelint language server: ${msg}`);
			throw err;
		}
	}
}
