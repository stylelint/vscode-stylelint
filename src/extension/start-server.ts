import path from 'path';
import process from 'process';
import { createConnection, ProposedFeatures } from 'vscode-languageserver/node';
import { StylelintLanguageServer, modules, createLogger } from '../server/index';

const connection = createConnection(ProposedFeatures.all);

const { NODE_ENV } = process.env;

const logger =
	NODE_ENV === 'development'
		? createLogger(connection, 'debug', path.join(__dirname, '../stylelint-language-server.log'))
		: createLogger(connection, 'info');

const server = new StylelintLanguageServer({
	connection,
	logger,
	modules: Object.values(modules),
});

server.start();
