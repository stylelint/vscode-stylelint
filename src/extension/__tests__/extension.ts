jest.doMock('vscode', () => ({ workspace: {}, window: {}, commands: {} }), {
	virtual: true,
});
jest.mock('vscode-languageclient/node', () => ({
	LanguageClient: jest.fn(),
	SettingMonitor: jest.fn(),
	ExecuteCommandRequest: {
		type: 'executeCommand',
	},
}));

import { EventEmitter } from 'events';
import vscode, { window } from 'vscode';
import { LanguageClient, SettingMonitor, NodeModule } from 'vscode-languageclient/node';
import {
	DidRegisterDocumentFormattingEditProviderNotificationParams,
	Notification,
} from '../../server/index';
import { activate, deactivate } from '../extension';
import { ApiEvent } from '../types';

const mockVSCode = vscode as jest.Mocked<typeof vscode>;
const mockLanguageClient = LanguageClient as jest.MockedClass<typeof LanguageClient>;
const mockSettingMonitor = SettingMonitor as jest.MockedClass<typeof SettingMonitor>;

const mockWorkspace = {
	createFileSystemWatcher: jest.fn(),
};

const mockCommands = {
	registerCommand: jest.fn(),
};

const mockTextEditor = {
	document: {
		uri: {
			toString: () => 'file:///path/to/file.ts',
		},
		version: 1,
	},
} as vscode.TextEditor;

const mockWindow = {
	activeTextEditor: mockTextEditor,
	showErrorMessage: jest.fn(),
};

mockVSCode.workspace = mockWorkspace as unknown as typeof vscode.workspace;
mockVSCode.commands = mockCommands as unknown as typeof vscode.commands;
mockVSCode.window = mockWindow as unknown as typeof vscode.window;

const start = jest.fn();
const stop = jest.fn();
const onNotification = jest.fn();
const sendRequest = jest.fn();
const settingMonitorStart = jest.fn();

const mockExtensionContext = {
	subscriptions: [],
} as unknown as vscode.ExtensionContext;

type ModuleServerOptions = { run: NodeModule; debug: NodeModule };

const isModuleOptions = (options: unknown): options is ModuleServerOptions =>
	Boolean((options as ModuleServerOptions)?.run?.module);

const stripPaths = <T extends unknown[]>(params: T): T => {
	if (Array.isArray(params)) {
		const serverOptions = params[1];

		if (isModuleOptions(serverOptions)) {
			serverOptions.run.module = 'mock-path';
			serverOptions.debug.module = 'mock-path';
		}
	}

	return params;
};

describe('Extension entry point', () => {
	beforeEach(() => {
		jest.resetAllMocks();

		mockVSCode.window.activeTextEditor = undefined;

		start.mockImplementation(async () => undefined);
		sendRequest.mockImplementation(async () => undefined);

		mockLanguageClient.mockReturnValue({
			onNotification,
			sendRequest,
			start,
			stop,
		} as unknown as LanguageClient);

		mockSettingMonitor.mockReturnValue({
			start: settingMonitorStart,
		} as unknown as SettingMonitor);

		(mockExtensionContext as { subscriptions: unknown[] }).subscriptions = [];
	});

	it('should provide a public API', async () => {
		const api = await activate(mockExtensionContext);

		expect(api).toBeInstanceOf(EventEmitter);
	});

	it('should create a language client', async () => {
		await activate(mockExtensionContext);

		expect(mockLanguageClient).toHaveBeenCalled();
		expect(stripPaths(mockLanguageClient.mock.calls[0])).toMatchSnapshot();
	});

	it('should watch for changes to Stylelint configuration files', async () => {
		await activate(mockExtensionContext);

		expect(mockWorkspace.createFileSystemWatcher).toHaveBeenCalledTimes(2);
		expect(mockWorkspace.createFileSystemWatcher.mock.calls[0]).toEqual([
			'**/.stylelintrc{,.js,.json,.yaml,.yml}',
		]);
		expect(mockWorkspace.createFileSystemWatcher.mock.calls[1]).toEqual([
			'**/{stylelint.config.js,.stylelintignore}',
		]);
	});

	it('should register an auto-fix command', async () => {
		const disposable = { dispose: () => undefined };

		mockCommands.registerCommand.mockReturnValueOnce(disposable);

		await activate(mockExtensionContext);

		const { subscriptions } = mockExtensionContext;

		expect(mockCommands.registerCommand).toHaveBeenCalled();
		// cspell:disable
		expect(mockCommands.registerCommand.mock.calls[0]).toMatchObject([
			'stylelint.executeAutofix',
			expect.any(Function),
		]);
		// cspell:enable
		expect(subscriptions).toContain(disposable);
	});

	it('with an active text editor, should send auto-fix commands to the language server', async () => {
		window.activeTextEditor = mockTextEditor;

		await activate(mockExtensionContext);

		await mockCommands.registerCommand.mock.calls[0][1](undefined);

		expect(sendRequest).toHaveBeenCalledTimes(1);
		expect(sendRequest.mock.calls[0]).toMatchSnapshot();
		expect(mockWindow.showErrorMessage).not.toHaveBeenCalled();
	});

	it('without an active text editor, should not send auto-fix commands to the language server', async () => {
		window.activeTextEditor = undefined;

		await activate(mockExtensionContext);

		await mockCommands.registerCommand.mock.calls[0][1](undefined);

		expect(sendRequest).not.toHaveBeenCalled();
		expect(mockWindow.showErrorMessage).not.toHaveBeenCalled();
	});

	it('should show an error message if sending the command request fails', async () => {
		window.activeTextEditor = mockTextEditor;

		await activate(mockExtensionContext);

		sendRequest.mockRejectedValueOnce(new Error('mock error'));

		await mockCommands.registerCommand.mock.calls[0][1](undefined);

		expect(sendRequest).toHaveBeenCalledTimes(1);
		expect(mockWindow.showErrorMessage).toHaveBeenCalledTimes(1);
		expect(mockWindow.showErrorMessage.mock.calls[0]).toEqual([
			'Failed to apply Stylelint fixes to the document. Please consider opening an issue with steps to reproduce.',
		]);
	});

	it('should register a restart server command', async () => {
		const disposable = { dispose: () => undefined };

		mockCommands.registerCommand.mockReturnValue(disposable);

		await activate(mockExtensionContext);

		const { subscriptions } = mockExtensionContext;

		expect(mockCommands.registerCommand).toHaveBeenCalled();
		expect(mockCommands.registerCommand.mock.calls[1]).toMatchObject([
			'stylelint.restart',
			expect.any(Function),
		]);
		expect(subscriptions).toContain(disposable);
	});

	it('should restart the language server', async () => {
		await activate(mockExtensionContext);

		await mockCommands.registerCommand.mock.calls[1][1]();

		expect(stop).toHaveBeenCalled();
		expect(start).toHaveBeenCalled();
	});

	it('should monitor settings', async () => {
		const disposable = { dispose: () => undefined };

		settingMonitorStart.mockReturnValueOnce(disposable);

		await activate(mockExtensionContext);

		const { subscriptions } = mockExtensionContext;

		const mockLanguageClientInstance = mockLanguageClient.mock.results[0].value;

		expect(mockSettingMonitor).toHaveBeenCalled();
		expect(mockSettingMonitor.mock.calls[0][0]).toBe(mockLanguageClientInstance);
		expect(mockSettingMonitor.mock.calls[0][1]).toBe('stylelint.enable');
		expect(settingMonitorStart).toHaveBeenCalled();
		expect(subscriptions).toContain(disposable);
	});

	it('should listen for the DidRegisterCodeActionRequestHandler notification', async () => {
		await activate(mockExtensionContext);

		await new Promise((resolve) => setImmediate(resolve));

		expect(onNotification).toHaveBeenCalled();
		expect(onNotification.mock.calls[0][0]).toBe(Notification.DidRegisterCodeActionRequestHandler);
		expect(onNotification.mock.calls[0][1]).toBeInstanceOf(Function);
	});

	it('should set codeActionReady to true when the DidRegisterCodeActionRequestHandler notification is received', async () => {
		const api = await activate(mockExtensionContext);

		await new Promise((resolve) => setImmediate(resolve));

		onNotification.mock.calls[0][1]();

		expect(api.codeActionReady).toBe(true);
	});

	it('should listen for the DidRegisterDocumentFormattingEditProvider notification', async () => {
		await activate(mockExtensionContext);

		await new Promise((resolve) => setImmediate(resolve));

		expect(onNotification).toHaveBeenCalled();
		expect(onNotification.mock.calls[1][0]).toBe(
			Notification.DidRegisterDocumentFormattingEditProvider,
		);
		expect(onNotification.mock.calls[1][1]).toBeInstanceOf(Function);
	});

	it('should emit the DidRegisterDocumentFormattingEditProvider event when the DidRegisterDocumentFormattingEditProvider notification is received', async () => {
		const api = await activate(mockExtensionContext);

		const promise = new Promise<DidRegisterDocumentFormattingEditProviderNotificationParams>(
			(resolve) => {
				api.on(ApiEvent.DidRegisterDocumentFormattingEditProvider, resolve);
			},
		);

		const params: DidRegisterDocumentFormattingEditProviderNotificationParams = {
			uri: 'file:///foo.css',
			options: {
				documentSelector: [{ scheme: 'file', pattern: '/foo.css' }],
			},
		};

		await new Promise((resolve) => setImmediate(resolve));

		onNotification.mock.calls[1][1](params);

		await expect(promise).resolves.toStrictEqual(params);
	});

	it('should listen for the DidResetConfiguration notification', async () => {
		await activate(mockExtensionContext);

		await new Promise((resolve) => setImmediate(resolve));

		expect(onNotification).toHaveBeenCalled();
		expect(onNotification.mock.calls[2][0]).toBe(Notification.DidResetConfiguration);
		expect(onNotification.mock.calls[2][1]).toBeInstanceOf(Function);
	});

	it('should emit the DidResetConfiguration event when the DidResetConfiguration notification is received', async () => {
		const api = await activate(mockExtensionContext);

		const promise = new Promise<void>((resolve) => {
			api.on(ApiEvent.DidResetConfiguration, resolve);
		});

		await new Promise((resolve) => setImmediate(resolve));

		onNotification.mock.calls[2][1]();

		await expect(promise).resolves.toBeUndefined();
	});

	it('should show an error message if registering notifications fails', async () => {
		onNotification.mockImplementation((): void => {
			throw new Error('Problem!');
		});

		await activate(mockExtensionContext);

		await new Promise((resolve) => setImmediate(resolve));

		onNotification.mockImplementation((): void => {
			throw 'String problem!'; // eslint-disable-line @typescript-eslint/only-throw-error
		});

		await activate(mockExtensionContext);

		await new Promise((resolve) => setImmediate(resolve));

		expect(mockWindow.showErrorMessage).toHaveBeenCalledTimes(2);
		expect(mockWindow.showErrorMessage.mock.calls[0]).toEqual(['Stylelint: Problem!']);
		expect(mockWindow.showErrorMessage.mock.calls[1]).toEqual(['Stylelint: String problem!']);
	});

	it('should show an error message if restarting the language server fails', async () => {
		await activate(mockExtensionContext);

		await new Promise((resolve) => setImmediate(resolve));

		start.mockImplementation((): void => {
			throw new Error('Problem!');
		});

		await mockCommands.registerCommand.mock.calls[1][1]();

		start.mockImplementation((): void => {
			throw 'String problem!'; // eslint-disable-line @typescript-eslint/only-throw-error
		});

		await mockCommands.registerCommand.mock.calls[1][1]();

		expect(mockWindow.showErrorMessage).toHaveBeenCalledTimes(2);
		expect(mockWindow.showErrorMessage.mock.calls[0]).toEqual(['Stylelint: Problem!']);
		expect(mockWindow.showErrorMessage.mock.calls[1]).toEqual(['Stylelint: String problem!']);
	});

	it('should stop language client on deactivate', async () => {
		await activate(mockExtensionContext);
		await deactivate();

		expect(stop).toHaveBeenCalledTimes(1);
	});

	it('should show error message when deactivate fails to stop the client', async () => {
		stop.mockImplementation(() => {
			throw new Error('foo');
		});

		await activate(mockExtensionContext);

		try {
			await deactivate();
		} catch {
			/* empty */
		}

		expect(window.showErrorMessage).toHaveBeenCalledTimes(1);
		expect(window.showErrorMessage).toHaveBeenCalledWith(
			'error stopping stylelint language server: foo',
		);
	});

	it('should show unknown error message when deactivate fails to stop the client', async () => {
		stop.mockImplementation(() => {
			throw undefined; // eslint-disable-line @typescript-eslint/only-throw-error
		});

		await activate(mockExtensionContext);

		try {
			await deactivate();
		} catch {
			/* empty */
		}

		expect(window.showErrorMessage).toHaveBeenCalledTimes(1);
		expect(window.showErrorMessage).toHaveBeenCalledWith(
			'error stopping stylelint language server: unknown',
		);
	});
});
