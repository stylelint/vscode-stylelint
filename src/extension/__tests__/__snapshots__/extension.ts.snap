// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extension entry point should create a language client 1`] = `
[
  "Stylelint",
  {
    "debug": {
      "module": "mock-path",
      "options": {
        "execArgv": [
          "--nolazy",
          "--inspect=6004",
        ],
      },
    },
    "run": {
      "module": "mock-path",
    },
  },
  {
    "diagnosticCollectionName": "Stylelint",
    "documentSelector": [
      {
        "scheme": "file",
      },
      {
        "scheme": "untitled",
      },
    ],
    "synchronize": {
      "fileEvents": [
        undefined,
        undefined,
      ],
    },
  },
]
`;

exports[`Extension entry point with an active text editor, should send auto-fix commands to the language server 1`] = `
[
  "executeCommand",
  {
    "arguments": [
      {
        "uri": "file:///path/to/file.ts",
        "version": 1,
      },
    ],
    "command": "stylelint.applyAutoFix",
  },
]
`;
